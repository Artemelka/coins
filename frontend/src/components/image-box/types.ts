import { SyntheticEvent } from 'react';

export type ImageBoxClickEvent = {
    id: number;
    event: SyntheticEvent;
};