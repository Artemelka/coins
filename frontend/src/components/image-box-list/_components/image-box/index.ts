export { ImageBox } from './image-box';
export type { ImageBoxClickEvent } from './types';