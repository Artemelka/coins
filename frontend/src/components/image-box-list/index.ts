export { ImageBoxList } from './image-box-list';