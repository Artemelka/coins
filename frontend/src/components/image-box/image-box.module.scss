.root {
  border-radius: 12px;
  border: 1px solid #000;
  padding: 30px;
  width: 340px;
  height: 470px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  flex-direction: column;
}

.image {
  display: block;
  max-width: 100%;
  height: auto;
}

.title {
  font-size: 24px;
  margin: 0;
}