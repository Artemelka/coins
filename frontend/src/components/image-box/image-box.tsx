import React, { FC, memo } from 'react';
import styles from './image-box.module.scss';

type IProps = {
    imageUri: string;
    title: string;
};

const ImageBoxComponent: FC<IProps> = ({ imageUri, title }) => {
    return (
        <div className={styles.root}>
            <img className={styles.image} src={imageUri} alt={title}/>
            <p className={styles.title}>{title}</p>
        </div>
    );
}

export const ImageBox = memo(ImageBoxComponent);