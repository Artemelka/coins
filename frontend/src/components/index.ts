export { ThemeSwitcher } from './theme-switcher';
export { ImageBoxList } from './image-box-list';
export { LazyImage } from './lazy-image';
export { Page } from './page';

export type { ImageBoxClickEvent } from './image-box-list/_components/image-box';