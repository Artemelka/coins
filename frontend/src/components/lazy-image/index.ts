export { LazyImage } from './lazy-image';