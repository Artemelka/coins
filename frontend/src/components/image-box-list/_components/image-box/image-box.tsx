import React, { FC, memo, SyntheticEvent, useCallback } from 'react';
import { LazyImage } from '@/components';
import { ImageBoxClickEvent } from './types';
import styles from './image-box.module.scss';

type ImageBoxProps = {
    id: number;
    imageUri: string;
    onClick: (clickEvent: ImageBoxClickEvent) => void;
    onImageLoad?: () => void;
    title: string;
    hideSkeleton?: boolean;
};

const ImageBoxComponent: FC<ImageBoxProps> = ({
    id,
    imageUri,
    onClick,
    title,
    onImageLoad,
    hideSkeleton,
}) => {
    const handleClick = useCallback((event: SyntheticEvent) => {
        onClick({ id, event });
    }, [id, onClick]);

    return (
        <button className={styles.root} type="button" onClick={handleClick}>
            <div className={styles.wrapper}>
                <LazyImage
                  src={imageUri}
                  alt={title}
                  onLoad={onImageLoad}
                  remoteControl={Boolean(onImageLoad)}
                  hideSkeleton={hideSkeleton}
                />
            </div>
            <p className={styles.title}>{title}</p>
        </button>
    );
}

export const ImageBox = memo(ImageBoxComponent);