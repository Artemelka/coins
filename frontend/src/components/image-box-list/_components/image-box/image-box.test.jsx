import React from 'react';
import { shallow } from 'enzyme';
import { LazyImage } from '@/components';
import { noop } from '@/utils';
import { ImageBox } from './image-box';

function getComponent(props = {}) {
  return (
    <ImageBox
      id={12}
      imageUri={'testUrl'}
      onClick={noop}
      title={'testTitle'}
      {...props}
    />
  );
}

describe('ImageBox', () => {
  describe('render', () => {
    test('default', () => {
      const component = shallow(getComponent());

      expect(component).toMatchSnapshot();
    });

    test('with all props', () => {
      const component = shallow(getComponent({ onImageLoad: noop, hideSkeleton: true }));

      expect(component).toMatchSnapshot();
    });
  });

  test('onClick should be call with id props', () => {
    const testOnClick = jest.fn();
    const id = 15;
    const testEvent = { preventDefault: noop };
    const component = shallow(getComponent({ id, onClick: testOnClick }));

    component.simulate('click', testEvent);

    expect(testOnClick).toBeCalledWith({ id, event: testEvent })
  });

  test('title should contains title prop', () => {
    const title = 'test';
    const component = shallow(getComponent({ title }));

    expect(component.find('.title').text()).toBe(title);
  });

  test('ImageBox should be button', () => {
    const component = shallow(getComponent());

    expect(component.find('.root').type()).toBe('button')
  });

  test('LazyImage props remoteControl=false if component don`t have onImageLoad', () => {
    const component = shallow(getComponent());

    expect(component.find(LazyImage).props().remoteControl).toBe(false);
  });

  test('LazyImage props remoteControl=true if component have onImageLoad', () => {
    const component = shallow(getComponent({ onImageLoad: noop }));

    expect(component.find(LazyImage).props().remoteControl).toBe(true);
  });

  test('LazyImage props hideSkeleton should be equal hideSkeleton prop', () => {
    const component = shallow(getComponent({ hideSkeleton: true }));

    expect(component.find(LazyImage).props().hideSkeleton).toBe(true);
  });

});