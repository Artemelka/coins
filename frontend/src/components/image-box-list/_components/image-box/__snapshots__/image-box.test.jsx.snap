// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageBox render default 1`] = `
<button
  className="root"
  onClick={[Function]}
  type="button"
>
  <div
    className="wrapper"
  >
    <Memo(LazyImageComponent)
      alt="testTitle"
      remoteControl={false}
      src="testUrl"
    />
  </div>
  <p
    className="title"
  >
    testTitle
  </p>
</button>
`;

exports[`ImageBox render with all props 1`] = `
<button
  className="root"
  onClick={[Function]}
  type="button"
>
  <div
    className="wrapper"
  >
    <Memo(LazyImageComponent)
      alt="testTitle"
      hideSkeleton={true}
      onLoad={[Function]}
      remoteControl={true}
      src="testUrl"
    />
  </div>
  <p
    className="title"
  >
    testTitle
  </p>
</button>
`;
